### ISSUE TITLE
code: tsdb transactions failing due to inline writes

### ISSUE THREAD

**Issue by mberhault** [cockroachdb/cockroach#10358]
Gamma (sha 5bc7bf1)
```
E161101 02:47:55.305985 65 storage/queue.go:558  [n5,timeSeriesMaintenance] on [n5,s5,r573/3:/System/tsd/cr.node.gossip.by…]: "/System/tsd/cr.node.gossip.bytes.received-1h/1/10s/2016-09-27T03:00:00Z": inline writes not allowed within transactions
```

This is constantly happening. Not sure if this is a misuse by the metrics code, or a deeper problem.

**Comment by mrtracy** [comment]
I don't have any theories on this. If this always happens in the TimeSeriesMaintenance queue, it has to be related to the deletion of old time series - this is definitely backed up by the age of the key it's trying to write. However, the system very explicitly doesn't use a transaction.

This functionality is tested using a TestServer; we start a test server, add and verify metrics, then force the queue to run and verify that metrics were deleted. That suggests that it's not that simple.

Investigating. Does anyone know offhand if there is anywhere in the DB/client stack where a transaction could be created automatically?

**Comment by tschottdorf** [comment]
If you're going through `TxnCoordSender`, you may be forced into a transaction when you span ranges. I bet this is what's happening and I *think* I pointed that out somewhere, but might've been on the RFC. Or not, but in any case not specifying a transaction does not guarantee you won't wind up in one.

**Comment by mrtracy** [comment]
Apparently this comment was added to the RFC, but I suppose it was not clear to me that the transaction was actually *forced* by the system.  It was also mentioned in the context of mixing Scans and Rollups - this apparently can happen on a batch that has only a single DeleteRange command. Okay, so there are two possible solutions here:

1. Set the batch's ReadConsistency to Inconsistent; this appears to prevent the "retry in transaction" effect from happening, but I don't know if it would have other side effects on the operation.
2. Create a new "DeleteRangeInline" op, separate from DeleteRange which does not have the isTxn or isTxnWrite property.

**Comment by tschottdorf** [comment]
Option 1) is messy; you shouldn't even be able to run Raft commands inconsistently (if you found that that works, we need to add an error check, but you would have to color me surprised).

I'd go with option 2), a new command which has `isWrite | isRange`, though I would prefer making it explicitly about time series (easy enough to rename if it becomes useful elsewhere, all things being internal).

**Comment by tschottdorf** [comment]
edit: and probably also `isAlone`.

**Comment by benmargolin** [comment]
I am seeing this a lot on my server, 
Build Tag:   beta-20161103
Build Time:  2016/11/03 15:03:54
Platform:    linux amd64
Go Version:  go1.7.3
C Compiler:  gcc 4.9.2
is this fix (a) pretty sure to really be fixed and (b) in that build? Seems to have been submitted pretty soon before the date on this build. Is there a new build available?

**Comment by a-robinson** [comment]
Unfortunately, the commit that fixed this *just* missed that release.

Fortunately, we released beta-20161201 last week and it [includes a lot of fixes and improvements](https://www.cockroachlabs.com/docs/beta-20161201.html).

**Comment by benmargolin** [comment]
Thanks, I upgraded right before an important presentation (d'oh!) and I think so far it's been OK :)
